### DISCUSSION TITLE
Run mode line in seperate process / thread

### DISCUSSION THREAD

**Discussion by sdilts** [stumpwm/stumpwm#1111] [Ideas]
Currently, the mode line runs in the StumpWM process on  the main thread. This complicates several things including enabling mouse interaction and long running commands in the mode line can slow down StumpWM. Code in the mode line can also occasionally crash StumpWM itself like in #588 and #592.  The in-process mode line also makes #827, and #382 much worse than they should be.

This is a major change, but since there has been talk (and some progress on) using CLIM for the mode line, it might be time to start exploring whether this is feasible. There were efforts to put the mode line in a separate thread in #577, but that MR was ultimately reverted due to thread-unsafe code in users' mode line configurations.

There are at least two major challenges for implementing this:
1. There would need to be some kind of IPC between StumpWM proper and the modeline. The options I'm aware of would be to implement the protocol that i3  uses  to communicate that info to its bar, or try to use whatever standard is in X11.
2. We would need to find a way to make it backward compatible, or at least make it as easy to switch over as possible. The current mode line can pull anything out of StumpWM, but that probably won't be possible with a new  system.

Is this something worth considering or looking into, or is the potential to largely break backward compatibility a nonstarter?

**Comment by tychoish** [comment]
Hasn't it always been possible to use use an external mode-line/taskbar/tray implementation that would run in a separate process? I don't use such a thing, becuase having something more simple seems nice to me. 

Doing something in a thread as a starting point, seems like a good way to iterate toward the goal of having a more distinct mode-line. I don't particularly have an opinion about the protocol, as long as it feels reasonable to use from within stump.

The only bit of stumpwm-specific data that appears in my modeline is the index number of specific windows, which could be presented differently using some kind of popup/menu rather than needing to have it in the modeline. I think having a CL taskbar that "feels like stump" might be really swell in general as a way of pushing us (and stump) toward something more modular.

**Comment by PuercoPop** [comment]
> There would need to be some kind of IPC between StumpWM proper and the modeline.

We already have an RPC already exists. See [handle-stumpwm-commands in events.lisp]( https://github.com/stumpwm/stumpwm/blob/be183f40057aa0a4ea126a885c2bdbd8831c4d58/events.lisp#L287-L304). It works through x11 atoms. The gist of it is you set the property named "STUMPWM_COMMAND" and StumpWM will return the result in the property "STUMPWM_COMMAND_RESULT". Below is a short example of how you can use xprop to run the show-window-properties command.

```
puercopop@PuercoDesktop:~
$ xprop -root -format STUMPWM_COMMAND 8s -set STUMPWM_COMMAND "show-window-properties"
puercopop@PuercoDesktop:~
$ xprop -root | grep RESULT
STUMPWM_COMMAND_RESULT(STRING) = "class: st-256color\ninstance: st-256color\ntype: :NORMAL\nrole: \ntitle: tmux\n"
```

> This is a major change, but since there has been talk (and some progress on) using CLIM for the mode line, it might be time to start exploring whether this is feasible.

Although this is a major change because it is in a separate process it can be worked on as contrib which means it can be adopted on an as desired basis.


<hr />

On the topic of IPCs, if we decide that we need something more than what we currently have I rather we don't use i3. For better or worse, dbus is the IPC of the Linux desktop so I'd rather expose StumpWM as a dbus object. If we want something simpler I'm partial to Plan9's everything is a file-system approach that i3's precursor, wmii took. Just my 2c.

**Comment by szos** [comment]
Regarding the RPC based on x11 atoms: It seems that it doesnt work consistently with interactive commands. For example, running the command `"colon"` via xprop (or clx running in a separate image) brings up the prompt, wheras running the command `"windowlist"` in this way doesnt bring up a window, but does grab focus - using `C-n`/`C-p` + `RET` allows us to navigate the list and select despite it not being drawn. 

While I think the current system is good enough (at least for my needs), It should be consistent. This seems like a bug so I'll open up an issue regarding it.

**Comment by dmb2** [comment]
This would alleviate/fix a lot of issues related to the modeline and it would help modularize our codebase.  I'm supportive of any code contributed to this!
